src/Pos/Block/Slog/Logic.hs

```diff
@@ -303,8 +307,16 @@ slogRollbackBlocks blunds = do
         mconcat [forwardLinksBatch, inMainBatch]
 
 -- Common actions for rollback and apply.
-slogCommon :: MonadSlogApply ssc ctx m => LastBlkSlots -> m ()
+slogCommon
+    :: MonadSlogApply ssc ctx m
+    => LastBlkSlots
+    -> m ()
 slogCommon newLastSlots = do
     sanityCheckDB
     slogPutLastSlots newLastSlots
-    putSlottingData =<< GS.getSlottingData
+    -- We read from the database and write in the memory.
+    -- TODO(ks): This is unsafe! We don't have control over sequentiality and
+    -- use explicit indexing. It would be better if we have sorted EpochSlotData and
+    -- pass it without the index.
```

Comment:
> For example, `slogCommon` is called both from apply and rollback.

Yes, it's called from apply and rollback, that's why it's called `Common`. It contains functionality which should be used in both functions.

> So in effect no matter what happens, either an apply or rollback, it always reads SlottingData from database and writes into memory? That seems wrong.

Yes, that's what it does. Update system maintains slotting data. It can modify it (`usApplyBlocks`). After it's modified and written into DB we also want to put it into memory. However, there are two problems:
1. The first one was introduced by your PR: in case of rollback it can delete something, not add. I think the best way is to fetch this data from GState (you do it already), and then just use `writeTVar` to put whole `SlottingData` into memory.
2. The second one is that this is fetched before it's actually applied to DB 🤦‍♂️ 
I noticed it only now, but it was present for a long time :/

Also now I think it's more reasonable to have this logic in `usApplyBlocks` (and `usRollbackBlocks`). There we have `PollModifier` which in particular has slotting data modifications. It's ok to add `MonadSlotsData` to `USGlobalApplyMode` (but not to `MonadPoll`).

P. S. I think it can be done after this PR is merged because it took quite a lot of time already.